sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -301,12 +306,139 @@ case class GpuCaseWhen(
+              val whenBoolAdjusted: GpuColumnVector = cumulativePred match {
+                case Some(prev) =>
+                  withResource(prev.getBase.not()) { notPreviouslyTrue =>
+                    GpuColumnVector.from(
+                      whenBool.getBase.and(notPreviouslyTrue), DataTypes.BooleanType)
+                  }
+                case None =>
+                  whenBool.incRefCount()
+              }
+
+              withResource(whenBoolAdjusted) { _ =>
+                if (isAllTrue(whenBoolAdjusted)) {
+                  // if this WHEN predicate is true for all rows and no previous predicate has
+                  // been true then we can return immediately
+                  return GpuExpressionsUtils.columnarEvalToColumn(thenExpr, batch)
+                }
+                val thenValues = withResource(filterBatch(tbl,
+                    whenBoolAdjusted.getBase, colTypes)) {
+                  trueBatch => GpuExpressionsUtils.columnarEvalToColumn(thenExpr, trueBatch)
+                }
+                withResource(thenValues) { _ =>
+                  withResource(gather(whenBoolAdjusted.getBase, thenValues)) { gather =>
+                    currentValue = Some(currentValue match {
+                      case Some(v) =>
+                        withResource(v) { _ =>
+                          GpuColumnVector.from(whenBoolAdjusted.getBase.ifElse(gather, v.getBase),
+                            dataType)
+                        }
+                      case _ =>
+                        GpuColumnVector.from(gather.incRefCount(), dataType)
+                    })
+                  }
+
+                  cumulativePred = Some(cumulativePred match {
+                    case Some(prev) =>
+                      withResource(prev) { _ =>
+                        GpuColumnVector.from(whenBool.getBase.or(prev.getBase),
+                          DataTypes.BooleanType)
+                      }
+                    case _ =>
+                      whenBoolAdjusted.incRefCount()
+                  })
+
+                  if (isAllTrue(cumulativePred.get)) {
+                    // no need to process any more branches or the else condition
+                    return currentValue.get.incRefCount()
+                  }
+                }
+              }
+            }
+        }
+
+        // invert the cumulative predicate to get the ELSE predicate
+        withResource(cumulativePred.get.getBase.not()) { elsePredicate =>
+          // replace null predicates with true (because this is an inverted predicate)
+          val elsePredNoNulls = withResource(Scalar.fromBool(true)) { t =>
+            GpuColumnVector.from(elsePredicate.replaceNulls(t), DataTypes.BooleanType)
+          }
+          withResource(elsePredNoNulls) { _ =>
+            elseValue match {
+              case Some(expr) =>
+                if (isAllTrue(elsePredNoNulls)) {
```

Comment:
We can make this early-out even cheaper by moving it up earlier and avoiding the `elsePredicate` computation shenanigans.  We can check for `isAllFalse(cumulativePred)` to know that everything is the `ELSE` clause (or nulls if no `ELSE` clause).